### DISCUSSION TITLE
How to call eglot-reconnect in a function?

### DISCUSSION THREAD

**Discussion by fast-90** [joaotavora/eglot#1223] [Q&A] [answered]
I am using pyvenv and am trying to get eglot to reconnect after activating a virtual environment. I have tried the following:

```elisp
(use-package pyvenv
  :straight t
  :hook (pyvenv-post-activate-hooks . eglot-reconnect))
```
However I get an error saying `run-hooks: Wrong number of arguments: #<subr eglot-reconnect>, 0`. I can see in the function definition that `eglot-reconnect` takes three arguments, but I can't figure out what to pass it to get it to work (my elips skills are non existing). 

Anyone have an idea on how I can get eglot to reconnect after switching environments?

**Comment by joaotavora** [comment]
two arguments, not three.  And the second one is optional. Try to pass it the result of `(eglot-current-server)`.  Not sure it will work, it depends which buffer is active when pyvenv does its work, which might not be a buffer where `(eglot-current-server)` returns anything meaningful.  Anyway you may try it.

 I don't know how to craft this suggestion into `use-package` not even sure that is possible (but maybe someone else does).   Suggest you unlearn use-package when you can.

**Reply by fast-90** [reply, marked as answer]
Actually, after a few extra hours of trying different things I found the following solution working:

```elisp
(use-package pyvenv
  :straight t
  :hook (pyvenv-post-activate-hooks . (lambda() (call-interactively 'eglot-reconnect))))
```

Although I am intrigued by your comment regarding `use-package`. Could you elaborate your thoughts?

**Comment by joaotavora** [comment]
On Fri, May 12, 2023 at 11:42 AM fast-90 ***@***.***> wrote:

> Could you elaborate your thoughts?

OK, I guess.

Use-package is a macro that expands to other forms.  It doesn't
do any "work" by itself.  Unless one knows heart what use-package
is doing, it's harder to communicate the intent or the effect
of a user configuration, for example in a bug report or a
conversation such as this.

That's the main problem: it's opaque.  It's a DSL for where no
DSL was needed.  Lisp macros shouldn't be used for this IMHO
(I guess IMNSHO).

Can you guess the expansion of that form?  It should be something
like.

  (straight-install 'pyenv) ;; or whatever
  (add-hook 'pyvenv-post-activate-hooks (lambda () ...))

But is it?  No, u-p actually adds a "hook" suffix, so
the variable affected is pyvenv-post-activate-hooks-hook.
Is that what you meant? And should you use a dot there or
not?  And will it work if want to edebug your function
(I don't think so).

Of course you could learn all this by reading the manual on
use-package, but why should you fill your brain with this
redundant second language?  Why should you fill other
people's brains with it?

If you had chosen not to use use-package and used standard
elisp forms and function calls we would know exactly when
and how they are executed. And you could even step through
your function with edebug.

And if you had those two lines instead of u-p, are they any
worse than what you have now? I don't think so, IMO it's
better.  You want nice sectioning in your config file?
Just use a comment.  You want delayed execution? Use
with-eval-after-load.  You want to save up on some typing?
Use a local variable.

If you're interested, bring me examples of where you think
use-package is saving you time/typing and I can probably
rewrite them to be just as economical and more readable.

If you're not, don't worry.  This ship has sailed.  Everyone,
especially newcomers are using use-package, some thinking
that it is _the_ way to configure Emacs packages/libraries.
It's not.

João

**Reply by fast-90** [reply]
Thanks for sharing your view! I am indeed one of those newcomers (though I have no opinion on use-package at all), so hearing it from someone more experienced is helpful. As I am not well versed with elisp I simply stumbled upon use-package in other people's config. Once I find more time I will look into other options and try to learn more elisp, but for the time being I'm going to have to rely on the community for questions like these :).
